Automate running benchmarks for all engines

tellet-q opened the pull request:
Solves https://github.com/qdrant/vector-db-benchmark/issues/123

Run *-default experiment for each engine using random-100 or glove-25-angular dataset against single-node deployment. Note that OS and ES have dedicated single-node deployments with reduced memory to fit into default github runner.

The workflow triggers:
* on manual run
* on push to master
* on pull-request creation

tellet-q commented:
We can adjust the triggering on a job level by adding conditionals like this:
```   if: >
      !(
        startsWith(github.event.head_commit.modified, 'tests/') || 
        startsWith(github.event.head_commit.modified, 'scripts/') || 
        startsWith(github.event.head_commit.modified, 'monitoring/') ||
        contains(github.event.head_commit.modified, '.dockerignore') ||
        contains(github.event.head_commit.modified, '.gitignore') ||
        contains(github.event.head_commit.modified, '.pre-commit-config.yaml') ||
        contains(github.event.head_commit.modified, 'Dockerfile') ||
        contains(github.event.head_commit.modified, 'LICENSE') ||
        contains(github.event.head_commit.modified, 'README.md')
        contains(github.event.head_commit.modified, 'run_all_engines.sh')
        contains(github.event.head_commit.modified, 'sync_results.sh')
      )
```

This will NOT trigger the jobs if the changes ONLY include changes in the specified folders and files. For any other case the jobs will run. Unfortunately I'll have to configure each job like this, so it'll look a bit cumbersome.

KShivendu commented on tools/wait_for_green_status.sh:
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+# This scripts helps to wait for Opensearch|Elasticsearch status to become Green
+
+set -e
+
+SEARCH_CLUSTER_HOST=${1:-"localhost:9200"}
+
+# Wait until the search cluster host is available
+until $(curl --output /dev/null --silent --head --fail "$SEARCH_CLUSTER_HOST"); do
+    printf '.'
+    sleep 1 # Wait for 1 second
+done
+
+# Wait for ES/OS to start
+response=$(curl --write-out %{http_code} --silent --output /dev/null "$SEARCH_CLUSTER_HOST")
+
+until [ "$response" = "200" ]; do
+    response=$(curl --write-out %{http_code} --silent --output /dev/null "$SEARCH_CLUSTER_HOST")
+    >&2 echo "Search cluster is unavailable - sleep 1s"
+    sleep 1
+done
+
+# Wait for ES/OS status to turn Green
+health="$(curl -fsSL "$SEARCH_CLUSTER_HOST/_cat/health?h=status")"
+health="$(echo "$health" | sed -r 's/^[[:space:]]+|[[:space:]]+$//g')"
+
+until [ "$health" = 'green' ]; do
+    health="$(curl -fsSL "$SEARCH_CLUSTER_HOST/_cat/health?h=status")"
+    health="$(echo "$health" | sed -r 's/^[[:space:]]+|[[:space:]]+$//g')"
+    >&2 echo "Search cluster status is not green yet - sleep 1s"
```
Nice. 

I think it would be nicer if we put this waiting logic inside the python code itself. Will help users too. But that could be a task for another day =)

KShivendu replied:
Created https://github.com/qdrant/vector-db-benchmark/issues/135 :)

KShivendu reviewed (changes requested):
Great work! :raised_hands: 

I hope this will also expose UI (in `/actions`) to manually pick only one particular engine/dataset, right?

KShivendu commented:
> This will NOT trigger the jobs if the changes ONLY include changes in the specified folders and files. For any other case the jobs will run. Unfortunately I'll have to configure each job like this, so it'll look a bit cumbersome.

Interesting that we can do this.

@tellet-q Can we do something like this instead? 
```
      (
        startsWith(github.event.head_commit.modified, 'engine/{clients,server}/*<engine-name>*') || 
        startsWith(github.event.head_commit.modified, 'engine/base_client/')
      )
```

Where `<engine-name>` will vary for each job (engine)

tellet-q commented on .github/workflows/actions/run-engine-benchmark/action.yaml:
```diff
@@ -0,0 +1,38 @@
+    - name: Execution
+      shell: bash
+      run: |
+        flag="${{ inputs.engine }}"
+        if [[ "$flag" == *"elasticsearch"* || "$flag" == *"opensearch"* ]]; then
+          ./tools/wait_for_green_status.sh
+        fi
+        source $(poetry env info -p)/bin/activate
```
I'll update the variable name but not the poetry cmd.
You cannot use poetry shell in a non-interactive environment. When I use poetry shell I get something like below
```
Spawning shell within /home/runner/.cache/pypoetry/virtualenvs/vector-db-benchmark-qdAWVK0x-py3.10
(25, 'Inappropriate ioctl for device')
```

tellet-q commented:
> I hope this will also expose UI (in `/actions`) to manually pick only one particular engine/dataset, right?

Unfortunately, no. There are no changes in the UI.

tellet-q commented:
> @tellet-q Can we do something like this instead?
> 
> ```
>       (
>         startsWith(github.event.head_commit.modified, 'engine/{clients,server}/*<engine-name>*') || 
>         startsWith(github.event.head_commit.modified, 'engine/base_client/')
>       )
> ```
> 
> Where `<engine-name>` will vary for each job (engine)

Not exactly like this, but similar, yes. 
```
    if: >
      (
      startsWith(github.event.head_commit.modified, 'engine/clients/pgvector') ||
      startsWith(github.event.head_commit.modified, 'engine/servers/pgvector') ||
      startsWith(github.event.head_commit.modified, 'engine/base_client/')
      )
```
In this case the job will run ONLY if changes were made in specified folders.

KShivendu reviewed (approved):
If you want to try [this](https://github.com/qdrant/vector-db-benchmark/pull/134#issuecomment-2063697257), then please give it a shot. Otherwise, it lgtm anyways.

Let's see how this works when merged.
